### ISSUE TITLE
MimeType::detectByContent memory usage

### ISSUE THREAD

**Issue by Stadly** [thephpleague/flysystem#1172]
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | no
| Version     | 1.0.69

#### Summary

`\League\Flysystem\Util\MimeType::detectByContent` uses [`finfo_buffer`](https://www.php.net/manual/en/function.finfo-buffer.php), which is an extreme [memory hog](https://bugs.php.net/bug.php?id=79263).

Check out these two scripts I've created to illustrate the problem:
- [Memory usage of `finfo_file`](https://3v4l.org/M3YLG)
- [Memory usage of `finfo_buffer`](https://3v4l.org/LNNoV)

For a 16M file, `finfo_file` uses 20M memory. `finfo_buffer`, however, uses 244M!

This can lead to running out of memory when handling large files with `Flysystem`. 

#### How to reproduce

Store a large file (for example 1/4 of PHP's memory limit) using `Flysystem::write`. (The issues is not present with `Flysystem::writeStream`, as then `\League\Flysystem\Util\MimeType::detectByContent` does nothing).

#### Possible solutions

We could send only the first 1M of the content to `finfo_buffer`. This should be sufficient to determine the mime type:

```
return $finfo->buffer(substr($content, 0, 1024 * 1024)) ?: null;
```

We could write the content to a temporary file and use `finfo_file` instead:

```
$filename = tempnam('/tmp', 'finfo');
file_put_contents($filename, $content);
return $finfo->file($filename) ?: null;
```

We could of course also just wait for PHP to fix the issue. I've seen reports of the issue at least as far back as 2017, so I wouldn't hold my breath. Also, the issue would still be present when using current and past versions of PHP.

**Comment by frankdejonge** [comment]
Hi @Stadly, I've addressed this in PR #1183. The MimeType utility is a static interface in V1 (in V2 it'll just be constructor arguments. So for 1.1, I've added a dependency on the recently extracted league/mime-type-detection package. This will allow you to configure your own mime-type detection strategy. By default, it'll use the same strategy as now, but you can implement your own (or use one of the other ones provided in the package), which fixes your issue.

**Comment by frankdejonge** [comment]
This is now released, so this issue is fixed.

**Comment by VeryStrongFingers** [comment]
FYI it looks like there might be another (relevant, but unrelated?) issue resulting in the same memory eating outcome.
tl;dr: there is/was a bug in PHP core

**References**
    https://bugs.php.net/bug.php?id=78987
    https://bugs.astron.com/view.php?id=234
    https://bugs.php.net/bug.php?id=79263
